### ISSUE TITLE
Bug: Control-C on ipython (on extension code using cysignals) is broken

### ISSUE THREAD

**Issue by tornaria** [void-linux/void-packages#35712]
Correct behaviour (using python):
```
$ python
Python 3.10.2 (main, Jan 15 2022, 03:11:32) [GCC 10.2.1 20201203] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from cypari2.pari_instance import Pari
>>> a = Pari()(54853908712446157179434453567343931596301333824416758692208077566613224454501131209)
>>> a.factor() # takes ~ 1/2 hour
^CTraceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "cypari2/gen.pyx", line 4361, in cypari2.gen.Gen.factor
KeyboardInterrupt
>>> 
```
After running `a.factor()` I hit control-C (the statement takes ~ half hour to finish). Since this is running extension code (pari C library) python is not catching signals; that's why cypari uses cysignals which is supposed to catch the signal and raise the `KeyboardInterrupt`.

But when running the same code using ipython nothing happens:
```
$ ipython
Python 3.10.2 (main, Jan 15 2022, 03:11:32) [GCC 10.2.1 20201203]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.31.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from cypari2.pari_instance import Pari

In [2]: a = Pari()(54853908712446157179434453567343931596301333824416758692208077566613224454501131209)

In [3]: a.factor() # takes 1/2 hour
^C^C^C^C^C
```

This bug affects our sagemath which is using our system packages for ipython, cypari2, cysignals, etc. When using the vendored version of these packages in sagemath, this works ok:
```
$ ./sage -ipython
Python 3.10.2 (main, Jan 15 2022, 03:11:32) [GCC 10.2.1 20201203]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.29.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from cypari2.pari_instance import Pari

In [2]: a = Pari()(54853908712446157179434453567343931596301333824416758692208077566613224454501131209)

In [3]: a.factor()
^C---------------------------------------------------------------------------
KeyboardInterrupt                         Traceback (most recent call last)
<ipython-input-3-af0555404bb2> in <module>
----> 1 a.factor()

cypari2/gen.pyx in cypari2.gen.Gen.factor()

KeyboardInterrupt: 

In [4]: 
```
It's possible that there's something about our packages that is breaking signal handling.

For the record, the package versions are:
 - ipython: 7.31.0 in void, 7.29.0 in sagemath
 - cypari2: 2.1.2 in void, 2.1.2 in sagemath
 - cysignals: 1.11.2 in void, 1.10.3 in sagemath

Downgrading system ipython to 7.29.0 (which I happen to have in my cache) doesn't fix the issue. I will try to build cysignals-1.10.3 as a void package (since 1.11.2 is the first version to be shipped in void) and see if that changes anything.

**Comment by tornaria** [comment]
The issue is fixed by downgrading `python3-prompt_toolkit` to `3.0.24_1`, with every other package as currently in the repos (including the recent update to ipython 8.0.1).

The issue is present with prompt_toolkit 3.0.26_1 so it seems to be a regression between 3.0.24 and 3.0.26 (I don't have 3.0.25 in my cache).

**Comment by tornaria** [comment]
There's a strong smell in the [changelog](https://raw.githubusercontent.com/prompt-toolkit/python-prompt-toolkit/master/CHANGELOG) for 3.0.25 and 3.0.26:
```
3.0.26: 2022-01-27
------------------

Fixes:
- Fixes issue introduced in 3.0.25: Don't handle SIGINT on Windows.


3.0.25: 2022-01-27
------------------

Fixes:
- Use `DummyOutput` when `sys.stdout` is `None` and `DummyInput` when
  `sys.stdin` is `None`. This fixes an issue when the code runs on windows,
  using pythonw.exe and still tries to interact with the terminal.
- Correctly reset `Application._is_running` flag in case of exceptions in some
  situations.
- Handle SIGINT (when sent from another process) and allow binding it to a key
  binding. For prompt sessions, the behavior is now identical to pressing
  control-c.
- Increase the event loop `slow_duration_callback` by default to 0.5. This
  prevents printing warnings if rendering takes too long on slow systems.
```

**Comment by tornaria** [comment]
Reported upstream: https://github.com/prompt-toolkit/python-prompt-toolkit/issues/1576
